# **Bioinformatics Project - Computational Drug Discovery [Part 3] Descriptor Calculation and Dataset Preparation(Modified Version towards unknown smiles predcition purpose)**

Chanin Nantasenamat

[*'Data Professor' YouTube channel*](http://youtube.com/dataprofessor)

Modified by quantaosun@gmail.com


In **Part 3**, we will be calculating molecular descriptors that are essentially quantitative description of the compounds in the dataset. Finally, we will be preparing this into a dataset for subsequent model prediction in notebook 5.

---

## **Download PaDEL-Descriptor**

In [1]:
! wget https://github.com/dataprofessor/bioinformatics/raw/master/padel.zip
! wget https://github.com/dataprofessor/bioinformatics/raw/master/padel.sh

--2022-01-09 12:11:01--  https://github.com/dataprofessor/bioinformatics/raw/master/padel.zip
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/dataprofessor/bioinformatics/master/padel.zip [following]
--2022-01-09 12:11:01--  https://raw.githubusercontent.com/dataprofessor/bioinformatics/master/padel.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25768637 (25M) [application/zip]
Saving to: ‘padel.zip’

padel.zip           100%[===================>]  24.57M   163MB/s    in 0.2s    

2022-01-09 12:11:02 (163 MB/s) - ‘padel.zip’ saved [25768637/25768637]

--2022-01-09 12:11:02--  https://github

In [2]:
! unzip padel.zip

Archive:  padel.zip
   creating: PaDEL-Descriptor/
  inflating: __MACOSX/._PaDEL-Descriptor  
  inflating: PaDEL-Descriptor/MACCSFingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._MACCSFingerprinter.xml  
  inflating: PaDEL-Descriptor/AtomPairs2DFingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._AtomPairs2DFingerprinter.xml  
  inflating: PaDEL-Descriptor/EStateFingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._EStateFingerprinter.xml  
  inflating: PaDEL-Descriptor/Fingerprinter.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._Fingerprinter.xml  
  inflating: PaDEL-Descriptor/.DS_Store  
  inflating: __MACOSX/PaDEL-Descriptor/._.DS_Store  
   creating: PaDEL-Descriptor/license/
  inflating: __MACOSX/PaDEL-Descriptor/._license  
  inflating: PaDEL-Descriptor/KlekotaRothFingerprintCount.xml  
  inflating: __MACOSX/PaDEL-Descriptor/._KlekotaRothFingerprintCount.xml  
  inflating: PaDEL-Descriptor/config  
  inflating: __MACOSX/PaDEL-Descriptor/._config  
  inf

## **Load unknown smiles data, called "example.txt"**

Download the curated ChEMBL bioactivity data that has been pre-processed from Parts 1 and 2 of this Bioinformatics Project series. Here we will be using the **bioactivity_data_3class_pIC50.csv** file that essentially contain the pIC50 values that we will be using for building a regression model.

In [3]:
import pandas as pd

# We will convert a raw txt file with two lines of smiles to a standard dataframe by a csv intermediate

In [25]:
df3 = pd.read_csv('/content/sars-cov-2-local_bioactivity_data_preprocessed.csv')

In [26]:
df3

Unnamed: 0  ... bioactivity_class
0           0  ...            active

[1 rows x 4 columns]

Next the txt file will be save as a csv file, then import back again, to beautify the structure.

In [28]:
selection=['smiles']
df3[selection]

smiles
0  CN(N=C1)C=C1C2=CC=C(OC(F)=C3C4=CN(CC5=CC=CC=C5...

In [29]:
df4_selection = df3[selection]
df4_selection

smiles
0  CN(N=C1)C=C1C2=CC=C(OC(F)=C3C4=CN(CC5=CC=CC=C5...

In [30]:
df4_selection.to_csv('molecule.smi', sep= '\t', index=False, header=False)

In [31]:
! cat molecule.smi | head -5

CN(N=C1)C=C1C2=CC=C(OC(F)=C3C4=CN(CC5=CC=CC=C5)N=C4)C3=N2


In [32]:
! cat molecule.smi | wc -l

1


## **Calculate fingerprint descriptors, for the unknown smiles**


### **Calculate PaDEL descriptors**

In [33]:
! cat padel.sh

java -Xms1G -Xmx1G -Djava.awt.headless=true -jar ./PaDEL-Descriptor/PaDEL-Descriptor.jar -removesalt -standardizenitro -fingerprints -descriptortypes ./PaDEL-Descriptor/PubchemFingerprinter.xml -dir ./ -file descriptors_output.csv


In [34]:
! bash padel.sh

Processing AUTOGEN_molecule in molecule.smi (1/1). 
Descriptor calculation completed in 1.869 secs . Average speed: 1.87 s/mol.


In [35]:
! ls -l

total 25736
-rw-r--r-- 1 root root    13129 Jan  9 12:25 descriptors_output.csv
drwxr-xr-x 3 root root     4096 Jan  9 12:11 __MACOSX
-rw-r--r-- 1 root root       58 Jan  9 12:25 molecule.smi
drwxrwxr-x 4 root root     4096 May 30  2020 PaDEL-Descriptor
-rw-r--r-- 1 root root      231 Jan  9 12:11 padel.sh
-rw-r--r-- 1 root root 25768637 Jan  9 12:11 padel.zip
drwxr-xr-x 1 root root     4096 Dec 23 14:32 sample_data
-rw-r--r-- 1 root root   258771 Jan  9 12:23 sars-cov-2_bioactivity_data_2class_pIC50_pubchem_fp.csv
-rw-r--r-- 1 root root   258771 Jan  9 12:22 sars-cov-2_bioactivity_data_3class_pIC50_pubchem_fp.csv
-rw-r--r-- 1 root root    14767 Jan  9 12:11 sars-cov-2_bioactivity_data_processed_2classes_pIC50.csv
-rw-r--r-- 1 root root      109 Jan  9 12:24 sars-cov-2-local_bioactivity_data_preprocessed.csv


## **Preparing the X matrices**

### **X data matrix**

In [36]:
df3_X = pd.read_csv('descriptors_output.csv')

In [37]:
df3_X

Name  PubchemFP0  ...  PubchemFP879  PubchemFP880
0  AUTOGEN_molecule           1  ...             0             0

[1 rows x 882 columns]

In [38]:
df3_X = df3_X.drop(columns=['Name'])
df3_X

PubchemFP0  PubchemFP1  PubchemFP2  ...  PubchemFP878  PubchemFP879  PubchemFP880
0           1           1           1  ...             0             0             0

[1 rows x 881 columns]

 # Y variable

In [40]:
df3_Y = df3['bioactivity']
df3_Y

0    8.5
Name: bioactivity, dtype: float64

# Combine X and Y

In [41]:
dataset3 = pd.concat([df3_X,df3_Y], axis=1)
dataset3

PubchemFP0  PubchemFP1  PubchemFP2  ...  PubchemFP879  PubchemFP880  bioactivity
0           1           1           1  ...             0             0          8.5

[1 rows x 882 columns]

# You may have noticed that we now have a very long bit of 882 descriptor set, but when we built the model earlier, there are much less descriptors. This mismatch will be addressed later.

In [24]:
dataset3.to_csv('sars-cov-2-local_bioactivity_data_2class_pIC50_pubchem_fp.csv', index=False)

# **Let's download the CSV file to your local computer for the Part 3B (Model Building).**